core/node/tee_verifier_input_producer/src/lib.rs

```diff
@@ -107,19 +102,43 @@ impl TeeVerifierInputProducer {
             .await
             .context("expected miniblock to be executed and sealed")?;
 
-        // need a new connection in the next block
-        drop(connection);
+        let used_contract_hashes = l1_batch_header
+            .used_contract_hashes
+            .into_iter()
+            .map(u256_to_h256)
+            .collect();
+
+        // `get_factory_deps()` returns the bytecode in chunks of `Vec<[u8; 32]>`,
+        // but `fn store_factory_dep(&mut self, hash: H256, bytecode: Vec<u8>)` in `InMemoryStorage` wants flat byte vecs.
+        pub fn into_flattened<T, const N: usize>(data: Vec<[T; N]>) -> Vec<T> {
+            let (ptr, len, cap) = {
+                let mut me = std::mem::ManuallyDrop::new(data);
+                (me.as_mut_ptr(), me.len(), me.capacity())
+            };
+            let (new_len, new_cap) = {
+                // SAFETY:
+                // - `cap * N` cannot overflow because the allocation is already in
+                // the address space.
+                // - Each `[T; N]` has `N` valid elements, so there are `len * N`
+                // valid elements in the allocation.
+                unsafe { (len.unchecked_mul(N), cap.unchecked_mul(N)) }
+            };
+            // SAFETY:
+            // - `ptr` was allocated by `data`
+            // - `ptr` is well-aligned because `[T; N]` has the same alignment as `T`.
+            // - `new_cap` refers to the same sized allocation as `cap` because
+            // `new_cap * size_of::<T>()` == `cap * size_of::<[T; N]>()`
+            // - `len` <= `cap`, so `len * N` <= `cap * N`.
+            unsafe { Vec::<T>::from_raw_parts(ptr.cast(), new_len, new_cap) }
+        }
```

Comment:
Ugh, sorry to still be that guy, but 😅 
We do not _fully_ forbid `unsafe`, but we try to use it in places where it would have a noticeable impact.
Here we have a function that is invoked once per batch (a few minutes on the mainnet currently), so the benefits of having `unsafe` instead of creating a second `Vec` and recollecting are not really clear.

Reply:
Though it's a nice and well-documented function that uses unsafe, if I've seen one 👍

Reply:
copy and pasted from the std lib :)

Reply:
fixed, but it hurts my feelings :)